File: BOJ(Java 공부용)/2667.java
import java.io.*;
import java.util.ArrayList;
import java.util.Collections;

public class Main {
    static BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

    static int[] gox = {0, 1, -1, 0, 1, 1, -1, -1};
    static int[] goy = {1, 0, 0, -1, 1, -1, 1, -1};

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine());
        char[][] board = new char[n][n];
        boolean[][] visited = new boolean[n][n];
        for (int i = 0; i < n; i++) {
            String s = br.readLine();
            for (int j = 0; j < s.length(); j++) {
                if (s.charAt(j) == ' ') break;
                board[i][j] = s.charAt(j);
            }
        }

        ArrayList<Integer> ans = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                if (!visited[i][j] && board[i][j] == '1') {
                    ans.add(dfs(i, j, board, visited, n));
                }
            }
        }

        bw.write(ans.size() + "\n");
        Collections.sort(ans);
        for (int i = 0; i < ans.size(); i++) bw.write(ans.get(i) + "\n");
        bw.flush();
        bw.close();
    }

    private static int dfs(int x, int y, char[][] board, boolean[][] visited, int n) {
        int ret = 1;
        visited[x][y] = true;
        for (int i = 0; i < 4; i++) {
            int nx = x + gox[i], ny = y + goy[i];
            if (nx < 0 || nx >= n || ny < 0 || ny >= n || visited[nx][ny] || board[nx][ny] == '0') continue;
            ret += dfs(nx, ny, board, visited, n);
        }
        return ret;
    }
}

File: BOJ(Kotlin)/2667.kt
import java.io.BufferedReader
import java.io.BufferedWriter
import java.io.InputStreamReader
import java.io.OutputStreamWriter
import java.util.*

val gox = intArrayOf(-1,1,0,0)
val goy = intArrayOf(0,0,1,-1)
fun main(args: Array<String>) {
    val br = BufferedReader(InputStreamReader(System.`in`))
    val bw = BufferedWriter(OutputStreamWriter(System.out))
    val n = br.readLine().toInt()
    val board = Array(n){CharArray(n)}
    for (i in 0 until n) {
        val s = br.readLine()
        for (j in s.indices) {
            board[i][j] = s[j]
        }
    }
    val visited = Array(n){BooleanArray(n)}
    var cnt = 0
    val arr = arrayListOf<Int>()
    for (i in 0 until n) {
        for (j in 0 until n) {
            if (board[i][j] == '1' && !visited[i][j]) {
                cnt++
                arr.add(dfs(i,j,visited,n,board))
            }
        }
    }

    bw.write(cnt.toString()+"\n")
    arr.sort()
    for (value in arr) {
        bw.write(value.toString()+"\n")
    }
    bw.flush()
    bw.close()
}
fun dfs(x : Int, y : Int, visited : Array<BooleanArray>,n : Int,board : Array<CharArray>) : Int {

    visited[x][y] = true
    var ans = 1
    for (i in 0 until 4) {
        val nx = x+gox[i]
        val ny = y+goy[i]
        if (nx < 0 || nx>= n || ny<0 || ny>=n || visited[nx][ny] || board[nx][ny]=='0') continue
        ans += dfs(nx,ny,visited,n,board)
    }
    return ans
}
